contracts/hydro/src/testing_token_manager.rs

```diff
@@ -0,0 +1,405 @@
+            max_validator_shares_participating: lsm_token_info_provider
+                .max_validator_shares_participating,
+            hub_connection_id: lsm_token_info_provider.hub_connection_id.clone(),
+            hub_transfer_channel_id: lsm_token_info_provider.hub_transfer_channel_id.clone(),
+            icq_update_period: lsm_token_info_provider.icq_update_period,
+        },
+        TokenInfoProviderInstantiateMsg::LSM {
+            max_validator_shares_participating: lsm_token_info_provider
+                .max_validator_shares_participating,
+            hub_connection_id: lsm_token_info_provider.hub_connection_id.clone(),
+            hub_transfer_channel_id: lsm_token_info_provider.hub_transfer_channel_id.clone(),
+            icq_update_period: lsm_token_info_provider.icq_update_period,
+        },
+    ];
+
+    let res = instantiate(deps.as_mut(), env.clone(), info, msg.clone());
+    assert!(res.is_err());
+
+    assert!(res
+        .unwrap_err()
+        .to_string()
+        .to_lowercase()
+        .contains("only one lsm token info provider can be used."));
+}
+
+#[test]
+fn handle_token_info_provider_instantiate_reply_test() {
+    let (mut deps, env) = (mock_dependencies(no_op_grpc_query_mock()), mock_env());
+
+    let token_info_provider = TokenInfoProvider::Derivative(TokenInfoProviderDerivative {
+        contract: String::new(),
+        cache: HashMap::new(),
+    });
+
+    let contract_address = deps.api.addr_make("token_info_provider_1").to_string();
+
+    let mut encoded_data = vec![];
+    prost::encoding::string::encode(1, &contract_address, &mut encoded_data);
+
+    let reply_msg = Reply {
+        id: 0,
+        gas_used: 0,
+        payload: Binary::new(
+            to_json_vec(&ReplyPayload::InstantiateTokenInfoProvider(
+                token_info_provider,
+            ))
+            .unwrap(),
+        ),
+        // `data` field is deprecated, but it must be set because otherwise the compiler gives an error
+        #[allow(deprecated)]
+        result: SubMsgResult::Ok(SubMsgResponse {
+            events: vec![],
+            msg_responses: vec![MsgResponse {
+                type_url: String::new(), // not used in the test
+                value: Binary::from(encoded_data),
+            }],
+            data: None,
+        }),
+    };
+
+    let res = reply(deps.as_mut(), env, reply_msg);
+    assert!(res.is_ok());
+
+    let res = TOKEN_INFO_PROVIDERS.load(&deps.storage, contract_address.clone());
+    assert!(
+        res.is_ok(),
+        "expect token info provider not found in the store"
+    );
+
+    match res.unwrap() {
+        TokenInfoProvider::LSM(_) => {
+            panic!("expected derivative token info provider, found LSM one.")
+        }
+        TokenInfoProvider::Derivative(provider) => {
+            assert_eq!(provider.contract, contract_address);
+        }
+    }
+}
+
+#[test]
+fn add_remove_token_info_provider_test() {
```

Comment:
for all the new test cases I would encourage having brief summary comments above